src/connector/src/source/filesystem/opendal_source/opendal_reader.rs

```diff
@@ -181,23 +189,28 @@ impl<Src: OpendalSource> OpendalReader<Src> {
             }
             // note that the buffer contains the newline character
             debug_assert_eq!(n_read, line_buf.len());
-
-            // FIXME(rc): Here we have to use `offset + n_read`, i.e. the offset of the next line,
-            // as the *message offset*, because we check whether a file is finished by comparing the
-            // message offset with the file size in `FsFetchExecutor::into_stream`. However, we must
-            // understand that this message offset is not semantically consistent with the offset of
-            // other source connectors.
             let msg_offset = (offset + n_read).to_string();
-            batch.push(SourceMessage {
-                key: None,
-                payload: Some(std::mem::take(&mut line_buf).into_bytes()),
-                offset: msg_offset,
-                split_id: split.id(),
-                meta: SourceMeta::Empty,
-            });
+            if (object_name.ends_with(".gz") || object_name.ends_with(".gzip"))
+                && offset + n_read <= start_offset
```

Comment:
The logic here is incorrect because offset is assigned to `start_offset` in L168. A clearer approach is to return the reader as well as the read offset when constructing the reader. For gzip reader, the initial read offset will be `0` and for others the, the initial read offset will be `start_offset`